### ISSUE TITLE
Identity not being added to ssh-agent service (windows)

### ISSUE THREAD

**Issue by Exlipse7** [microsoft/vscode-remote-release#737]
- VSCode Version: 1.35.1 (user setup)
- Local OS Version: Windows_NT x64 10.0.18362
- Remote OS Version: CentOS Linux release 7.6.1810 (Core)
- Remote Extension/Connection Type: SSH (v0.42.2)

Steps to Reproduce:

1.  Start win10 ssh-agent service
2.  Open remote ssh connection
3.  Enter key password in vscode terminal
4.  Repeat entering key everytime
5.  Check ssh-add -l and see no keys listed

According to the Tips doc (https://code.visualstudio.com/docs/remote/troubleshooting#_setting-up-the-ssh-agent) vscode should automatically add my identity to the ssh-agent as long as the service is running.  This does not seem to be happening, it asks for a password every time and there's no keys listed in the agent.  If I manually add the key to the agent before opening the connection it is used and only results in the following warning: "warning: agent returned different signature type ssh-rsa (expected rsa-sha2-512)"

**Comment by strigona-worksight** [comment]
I'm running into the same issue. I think this is somewhat related to: https://github.com/PowerShell/Win32-OpenSSH/issues/1263

Which is pretty silly it's been half a year and still not truly fixed.

**Comment by roblourens** [comment]
I think that warning can be ignored right? But if we are failing to add the key to the agent it's probably because we can't find the agent, like if it's not on the path that vscode has or possibly if it was started after vscode was launched.

Could you share the full log from the "Remote - SSH" output channel?

![image](https://user-images.githubusercontent.com/323878/58114233-2c56fb00-7bac-11e9-8b26-23053f14763f.png)

**Comment by roblourens** [comment]
Also if you have already set `remote.SSH.showLoginTerminal`, which you probably have to when using a key with a passphrase and no agent, then we will not be trying to add the key to the agent.

**Comment by strigona-worksight** [comment]
`remote.SSH.showLoginTerminal` works around the issue. It's not the end of the world, but would be nice to use the agent to cut down on the number of times I have to enter the passphrase.

I'm working on getting work done right now and will try and get more logs to you later, but I didn't see much beyond the warning that @Exlipse7 mentioned at the end of his post.

**Comment by roblourens** [comment]
You should be able to add the key to the agent manually too. If vscode actually can't talk to the agent then that is a different issue.

**Comment by Exlipse7** [comment]
Alright, I didn't know that the `remote.SSH.showLoginTerminal` needed to be false.  I fixed that and it's still not working.  I've added the requested log below.  I verified that all VS Code processes were closed, verified the ssh-agent service was running and accessible from cmd.  Started VS Code and was prompted for password at the top of screen, entered it.  It looks like the server check went through but everything ground to a halt when it tried to reconnect (did not ask for password).  I closed VS Code and opened a local folder and verified that in the VS Code cmd terminal I could also access ssh-add (which worked as well).  Also, to reiterate, if I manually ssh-add my key before starting VS Code, then it uses the agent and I'm not asked for a password.  It seems that VS Code can talk to it, it's just the ssh-add that's failing.

Log:
```remote-ssh@0.42.2
SSH Resolver called for "ssh-remote+dev"
SSH Resolver called for host: dev
Setting up SSH remote "dev"
Using commit id "c7d83e57cd18f18026a8162d042843bda1bcf21f" and quality "stable" for server
Using SSH config file "c:\Users\user\.ssh\config"
Install and start server if needed
> 
> ]0;C:\Windows\SYSTEM32\cmd.exe
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> Enter passphrase for key 'C:\Users\user/.ssh/mykey': 
"install" wrote data to terminal: "*********"
> 
> 
> Found existing installation at /usr/user/.vscode-server/bin/c7d83e57cd18f18026a81
> 62d042843bda1bcf21f...
> 
> Found running server...
> 
> *
> * Reminder: You may only use this software with Visual Studio family products, 
> * as described in the license (https://go.microsoft.com/fwlink/?linkid=2077057)
> *
> 
> 
> dae433ae-c1ee-4a53-bc20-e064e0856f94==41817==
> 
"install" terminal command done
User entered passphrase
Passphrase was entered and ssh config entry was found - adding SSH key to agent
Executing add-ssh-key:
powershell -ExecutionPolicy Bypass C:\Users\user\AppData\Local\Temp\vscode_add_ssh_key_to_agent.ps1
"add-ssh-key" wrote data to terminal: "powershell -ExecutionPolicy Bypass C:\Users\user\AppData\Local\Temp\vscode_add_ssh_key_to_agent.ps1"
> 
> ]0;C:\Windows\SYSTEM32\cmd.exe
> Microsoft Windows [Version 10.0.18362.175]
> (c) 2019 Microsoft Corporation. All rights reserved.
> 
> c:\Users\user>
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> Microsoft Windows [Version 10.0.18362.175]
> (c) 2019 Microsoft Corporation. All rights reserved.
> 
> c:\Users\user>(powershell -ExecutionPolicy Bypass C:\Users\user\AppData\Local\
> Temp\vscode_add_ssh_key_to_agent.ps1) & exit /b 0
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> ]0;C:\Windows\SYSTEM32\cmd.exe - powershell  -ExecutionPolicy Bypass C:\Users\user\AppData\Local\Temp\vscode_add_ssh_key_to_agent.ps1
> Microsoft Windows [Version 10.0.18362.175]
> (c) 2019 Microsoft Corporation. All rights reserved.
> 
> c:\Users\user>(powershell -ExecutionPolicy Bypass C:\Users\user\AppData\Local\
> Temp\vscode_add_ssh_key_to_agent.ps1) & exit /b 0
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> PS: adding key to ssh-agent
> Microsoft Windows [Version 10.0.18362.175]
> (c) 2019 Microsoft Corporation. All rights reserved.
> 
> c:\Users\user>(powershell -ExecutionPolicy Bypass C:\Users\user\AppData\Local\
> Temp\vscode_add_ssh_key_to_agent.ps1) & exit /b 0
> PS: adding key to ssh-agent
> ssh-add
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> Microsoft Windows [Version 10.0.18362.175]
> (c) 2019 Microsoft Corporation. All rights reserved.
> 
> c:\Users\user>(powershell -ExecutionPolicy Bypass C:\Users\user\AppData\Local\
> Temp\vscode_add_ssh_key_to_agent.ps1) & exit /b 0
> PS: adding key to ssh-agent
> ssh-add
> 
> bcea7ca1-6636-4ffc-b8b1-c8c55796c4f6##5##
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
add-ssh-key failed
> ]0;C:\Windows\SYSTEM32\cmd.exe
add-ssh-key failed
> 
add-ssh-key failed
"add-ssh-key" terminal command done
Key with passphrase must be in ssh-agent
```

**Comment by strigona-worksight** [comment]
@roblourens 

- Setting `remote.SSH.showLoginTerminal` to True does work, with the "degraded experience" (paraphrasing the warning message).
- Setting `remote.SSH.showLoginTerminal` to False does not work. The error would be something similar to what was just posted by @Exlipse7 
- Setting `remote.SSH.showLoginTerminal` to False AND adding the key to the agent does not work. I don't have the full logs, but I did get this "warning: agent returned different signature type ssh-rsa (expected rsa-sha2-512)", which appears to be an issue with OpenSSH (see PowerShell/Win32-OpenSSH#1263)

System:
- Latest VS Code
- Extension is up to date
- Windows 10 x64 up to date
- OpenSSH Client via Windows Features

I have not tried patching the ssh-agent with the beta release from January.

**Comment by roblourens** [comment]
@Exlipse7 it looks like your key is not in the ssh config, right? Does it have a "default" name, like `.ssh/id_rsa`? Basically we are calling `ssh-add` with no argument, and it should find keys with default names. But if that's not the case, connecting wouldn't have worked in the first place anyway.

**Comment by Exlipse7** [comment]
@roblourens My key is not "id_rsa".  It is in the .ssh directory though and it is in my config file as `
IdentityFile ~/.ssh/myKey`.  I mean it asks for my password, so it's being loaded.  When the showTerminal was true everything worked fine, I just had to type my password a lot.  When I changed it to false, I entered my password the first time, and as you can see in the log, it connected and verified the vscode server.  It just failed to add to the agent and thus failed to connect on the 2nd trip.

I just looked through the log again, and I get it now.  You're not using any arguments, like not even the keyname.  So `ssh-add` won't do anything by itself, because that would only add `~/.ssh/id_rsa`.  You would have to use `ssh-add c:\Users\user\.ssh\mykey` (side note, openssh understands the `~/.ssh` path, but `ssh-add` does not).  I guess I expected it to add the key it used from the config file and not just blindly call `ssh-add`.

**Comment by strigona-worksight** [comment]
I can confirm, that my key is also in a different file and my `~/.ssh/config` file maps the remote host to the `IdentityFile`.

**Comment by roblourens** [comment]
>  I just looked through the log again, and I get it now. You're not using any arguments, like not even the keyname. 

That's right, so basically ssh can find your key but we can't find it when we parsed the ssh config for some reason. Can you share your ssh config so I can see if there's anything unusual about it causing us to fail at parsing it?

Another thing that might help, you can try the nightly version of the ssh extension (uninstall the stable version first) to see if any other changes have impacted this, if you have not tried that already.

**Comment by strigona-worksight** [comment]
Here's mine:

```
Host github.com
	User git
	IdentityFile ~/.ssh/github-worksight.key

Host lx1
	User simon
	IdentityFile ~/.ssh/lx1.key

Host lx2
	User simon
	IdentityFile ~/.ssh/lx1.key

Host weechat
	User simon
	IdentityFile ~/.ssh/weechat.key
	Port 9001

Host workstation
	User ad\simon
	IdentityFile ~/.ssh/st-ubt
	```

**Comment by Exlipse7** [comment]
My config file is fairly large now, but here's an example entry and the one I'm testing all of this on:
```
Host dev
HostName subdomain.sub.domain.com
IdentityFile ~/.ssh/myKey
User me
ForwardAgent yes
```
I also tried setting IdentityFile to `c:\Users\user\.ssh\myKey` as `ssh-add` doesn't seem to understand the `~/` syntax in CMD.  I haven't tried the nightly, so that will be my next move and I'll let you know.

**Comment by roblourens** [comment]
I have no idea. I guess if it used an `Include` to reference another file this could happen but that doesn't seem to be the case. 

> as ssh-add doesn't seem to understand the ~/ syntax in CMD

I was also just thinking that but if that was the case we would at least see the ~/ path to the key in the log.

**Comment by strigona-worksight** [comment]
Please keep open.

**Comment by thinkjrs** [comment]
> My config file is fairly large now, but here's an example entry and the one I'm testing all of this on:
> 
> ```
> Host dev
> HostName subdomain.sub.domain.com
> IdentityFile ~/.ssh/myKey
> User me
> ForwardAgent yes
> ```
> 
> I also tried setting IdentityFile to `c:\Users\user\.ssh\myKey` as `ssh-add` doesn't seem to understand the `~/` syntax in CMD. I haven't tried the nightly, so that will be my next move and I'll let you know.

I updated keys and ran into this issue on our Windows 10 machines + vscode remote ssh. 

This suggestion:
>
>I also tried setting IdentityFile to `c:\Users\user\.ssh\myKey` as `ssh-add` doesn't seem to understand the `~/` syntax in CMD

solved my problem! Vscode now properly references my key and uses the Windows-side agent 👍 

Thanks @Exlipse7!

**Comment by n-batalha** [comment]
I have a key in the default location (`~/.ssh/id_rsa`). Adding it as `IdentityFile C:\Users\nelson\.ssh\id_rsa` was not enough in terms of making VSCode storing the passphrase in the agent. 

But I can confirm `AddKeysToAgent yes` did it for me. 

This [doc](https://docs.microsoft.com/en-us/windows-server/administration/openssh/openssh_server_configuration#windows-configurations-in-sshd_config) states that the config should be in `%programdata%\ssh\sshd_config` but it's empty for me (this is with a plain install of ssh as a Windows 10 additional feature), as from what you say this value should be the default but seems it's not?

**Comment by Gregorek85** [comment]
> But I can confirm `AddKeysToAgent yes` did it for me.

Thank you, worked for me too

**Comment by meta-inf** [comment]
I tried all workarounds listed above and still can't get it to work.  Specifically I have added the following to my settings.json
```javascript
    "remote.SSH.path": "C:\\WINDOWS\\System32\\OpenSSH\\ssh.exe",
    "remote.SSH.useLocalServer": false,
    "remote.SSH.showLoginTerminal": false  // tried true, no difference
```
I have `AddKeysToAgent yes` in my `.ssh/config`, and my identity file has the default name.  I have `ssh-agent` running, and the local VSCode terminal can find the command.  If I `ssh-add` the key manually before launching VSCode, it can find the key; otherwise after typing the passphrase in VSCode, `ssh-add -l` still shows no keys. 

My platform:
* VSCode Version: 1.54.1 (system setup)
* Local OS Version: Windows_NT arm64 10.0.19042
* Remote Extension/Connection Type: SSH (v0.65.1)

With log level set to 2, Remote-SSH log has no mention of `ssh-agent` or `ssh-add`.

Any idea?

**Comment by escorciav** [comment]
Last year when I faced this, it was my bad.

I set up the equivalent of  `.ssh/config` in Windows manually using Git-CMD. Huge huge mistake. I deleted that. Used the standard and voila.

**Comment by meta-inf** [comment]
Thanks @escorciav , but not sure what you mean by "the standard".  I copied the config and ssh keys from WSL, but made sure the permissions are set up correctly (owned by the user, no other user has access).  Seems this is the only requirement for these files in the [troubleshooting guide](https://code.visualstudio.com/docs/remote/troubleshooting#_setting-up-the-ssh-agent).

<details>
<summary> the permissions </summary>

```
PS C:\Users\user\.ssh> get-acl config  # same for id_rsa

    Directory: C:\Users\user\.ssh

Path   Owner    Access
----   -----    ------
config <user>   NT AUTHORITY\SYSTEM Allow  FullControl...
```
</details>

**Comment by escorciav** [comment]
Sorry for not being clear. My advice, as it worked for me, is to isolate the issue.

- I focused on getting password-less to work through PowerShell 1st.

- Then, vscode.

By standard way, I mean forgetting that I'm using WSL, git-cmd, cywing, or any other bash terminal. The focus is establishing passwordless access from Windows to my server. Like you, all the permissions seemed to be OK. However, the way used to create the files came from my Linux experience. I did a backup of the folders, and create everything from scratch :wink: . Good luck!
